Changes

MisterDA opened the pull request:
Hi!
This PR introduces fixes and some changes:
- switches completely to the Windows Unicode API and handles conversion from OCaml strings to OS strings;
- doesn't use the `caml_winsvc_` prefix but only `winsvc_` as `caml_` is reserved by OCaml standard library;
- improves the error reporting by using system error messages;
- introduces a new exception to handle the case where a program isn't launched through the Windows Service manager;
- and adds small C improvements.
Let me know what you think!

MisterDA committed:
Use OCaml and Windows Unicode APIs

MisterDA committed:
Don't use the `caml_` prefix for C stubs

Using the `caml_` prefix here is unlikely to cause any problem at all
as the library really uses `caml_winsvc_`, however it's better not to
use the prefix at all.

MisterDA committed:
Mark unexported functions as static

MisterDA committed:
Use NULL for pointer comparison and simplify bool tests

MisterDA committed:
Retrieve error message with GetLastError and FormatMessage

dra27 commented on src/winsvc_stubs.c:
```diff
@@ -99,9 +99,9 @@ void service_main(DWORD argc, WCHAR **argv) {
   report_status(SERVICE_STOPPED, NO_ERROR, 2000);
 }
 
-#define raise_error(str) caml_raise_with_string(*caml_named_value("caml_service_exn"), str)
```
Fixes a bug as well - `"caml_service_exn"` here but `"caml_winsvc_exn"` above

dra27 commented on src/winsvc_stubs.c:
```diff
@@ -99,7 +100,31 @@ static void service_main(DWORD argc, WCHAR **argv) {
+
+  str = caml_stat_strdup_of_os(msg);
+  caml_raise_with_string(*caml_named_value("winsvc_exn"), str);
+  /* str leaks */
```
You can avoid this by using `caml_raise_with_arg`

dra27 commented on src/winsvc_stubs.c:
```diff
@@ -252,8 +252,12 @@ CAMLprim value winsvc_run(value v_name, value v_run, value v_stop) {
-    raise_error(L"StartServiceCtrlDispatcher", rc);
+  if (!result) {
+    if (rc == ERROR_FAILED_SERVICE_CONTROLLER_CONNECT)
+      caml_raise(*caml_named_value("winsvc_failed_service_controller_connect_exn"));
```
I think `caml_failwith` will do here?

dra27 reviewed (commented):
I'd just use `Failure _`, or possibly even `Not_found` - it's a uniquely useful error (as in you might do something other than terminate with it)

MisterDA committed:
Raise Failure if the program is run in console and not in service

If StartServiceCtrlDispatcher() returns
ERROR_FAILED_SERVICE_CONTROLLER_CONNECT, then the program is being run
as a console application rather than as a service. Not using
Winsvc.Error helps identifying this special case.

MisterDA commented:
Applied your suggestions, thanks!

toots commented:
This all looks great! My only remark is w.r.t. to the namespace. Liqiudsoap is built statically and I try to be on the most safe side I can with potentiasl symbol name clashes. I agree on the `caml_` prefix being reserved so I'll just go with `ocaml_winscv`, which is the conventions in our others bindings. Other than that, the changes look like great improvements. This code has been needing some love for a while. Thanks!
